File: Solutions/Unique_Paths_III/Unique_Paths_III.go
func uniquePathsIII(grid [][]int) int {
    zeros:=0
    sx,sy:=0,0
    for i:=0;i<len(grid);i++{
        for j:=0;j<len(grid[0]);j++{
            if grid[i][j]==0{
                zeros+=1
            } else if grid[i][j]==1{
                sx,sy=i,j
            }
        }
    }
    
    res:=0
    
    var dfs func(int,int,int)
    dfs=func(px,py,path int){
        directions:=[][]int{{0,1},{1,0},{0,-1},{-1,0}}
        
        for _,d:=range directions {
            newX,newY:=px+d[0],py+d[1]
            if newX>=0 && newX<len(grid) && newY>=0 && newY<len(grid[0]){
                if grid[newX][newY]==0 {
                    grid[newX][newY]=3
                    dfs(newX,newY,path+1)
                    //backtrack
                    grid[newX][newY]=0
                }
                if grid[newX][newY]==2 && path==zeros{
                    res+=1
                }
            }
        }
    }
    dfs(sx,sy,0)
    
    return res
}

//Another
func dfs(grid, visited *[][]int, x, y, steps, totalTraversable int, result *int){
    if x < 0 || y < 0 || x >= len((*grid)[0]) || y >= len(*grid) || (*grid)[y][x] == -1 || (*visited)[y][x] == -1{
        return
    }

    if steps + 1 == totalTraversable && (*grid)[y][x] == 2 {
        *result++
        return
    }else if (*grid)[y][x] == 2{
        return
    }
    
    steps++
    (*visited)[y][x] = -1
    dfs(grid, visited, x+1, y, steps, totalTraversable, result)
    dfs(grid, visited, x, y+1, steps, totalTraversable, result)
    dfs(grid, visited, x, y-1, steps, totalTraversable, result)
    dfs(grid, visited, x-1, y, steps, totalTraversable, result)
    // backtrack
    (*visited)[y][x] = 0
    steps--
}

func uniquePathsIII(grid [][]int) int {
    var startX, startY, result int
    m := len(grid)
    n := len(grid[0])
    totalTraversable := m*n
    
    visited := make([][]int, m)
    
    for i:=0; i<m;i++{
        visited[i] = make([]int, n)
        for j:=0; j<n;j++{
            if grid[i][j] == 1{
                startX = j
                startY = i
            }else if grid[i][j] == -1{
                totalTraversable--
            }
        }
    }
    dfs(&grid, &visited, startX, startY, 0, totalTraversable, &result)
    return result
}

File: Solutions/Unique_Paths_III/Unique_Paths_III.js
const dirs = [ [0, 1], [1, 0], [-1, 0], [0, -1] ];

const uniquePathsIII = grid => {
    let res = 0, traversable = 1, sx, sy;
    let m = grid.length, n = grid[0].length;
    
    for (let i = 0; i < m; i++)
        for (let j = 0; j < n; j++)
            if (grid[i][j] === 0)
                traversable++;
            else if (grid[i][j] === 1)
                sx = i, sy = j;

    dfs(sx, sy);
    
    return res;

    function dfs(r, c) {
        if (r < 0 || c < 0 || r >= m || c >= n || grid[r][c] < 0)
            return;

        if (grid[r][c] === 2)
            return res += !traversable;

        grid[r][c] = -2, traversable--;

        for (let [dR, dC] of dirs)
            dfs(r + dR, c + dC);

        grid[r][c] = 0, traversable++;
    }
};

//Fastest
/**
 * @param {number[][]} grid
 * @return {number}
 */
var uniquePathsIII = function(board) {
  let n = board.length;
  let m = board[0].length;
  let count = 0;
  
  // Pre-process board (count zeroes, find x,y start)
  let startX = 0
  let startY = 0;
  let zeroes = 1;
  for (let x = 0; x < n; x++) {
    for (let y = 0; y < m; y++) {
      if (board[x][y] === 0) zeroes += 1;
      if (board[x][y] === 1) {
        startX = x;
        startY = y;
      }
    } 
  }
  
  const dfs = (x, y, zeroes) => {
    if (x < 0 || x >= n) return;
    if (y < 0 || y >= m) return;
    if (board[x][y] === -1) return;
    
    //  if we reach the end with 0 zeroes left, add to path count
    if (board[x][y] === 2) {
      if (zeroes === 0) count += 1;
      return;
    }
    
    //  mark node as -1
    //  dfs in all 4 directions
    let orig = board[x][y];
    board[x][y] = -1;
    
    dfs(x + 1, y, zeroes - 1);
    dfs(x - 1, y, zeroes - 1);
    dfs(x, y + 1, zeroes - 1);
    dfs(x, y - 1, zeroes - 1);
    
    board[x][y] = orig;
  }
  
  // Recursively DFS from x,y of start
  dfs(startX, startY, zeroes);
  
  return count;
}